lib/commanded/event/handler.ex

```diff
@@ -1087,6 +1108,60 @@ defmodule Commanded.Event.Handler do
     %Handler{state | last_seen_event: event_number}
   end
 
+  defp snapshot_uuid(%Handler{} = state) do
+    %Handler{handler_module: handler_module, handler_name: handler_name} = state
+
+    inspect(handler_module) <> "-" <> inspect(handler_name)
+  end
+
+  defp read_state(%Handler{persistence: :permanent} = state) do
+    %Handler{application: application} = state
+
+    case EventStore.read_snapshot(application, snapshot_uuid(state)) do
+      {:ok, snapshot} ->
+        %SnapshotData{data: data, source_version: source_version} = snapshot
+
+        %Handler{
+          state
+          | handler_state: data,
+            last_seen_event: source_version
+        }
+
+      {:error, :snapshot_not_found} ->
+        state
+    end
+  end
+
+  defp read_state(state), do: state
+
+  defp persist_state(%Handler{persistence: :permanent} = state) do
```

Comment:
Snapshotting on every event smells like something that will severely limit the usability of this feature; there's a reason that Aggregate snapshotting is only between checkpoints. 

Not going to make a breaking point, if this is a useful feature as it is, that's fine, but I wonder how this thing would work when only snapshotting every X events or Y minutes...

Reply:
I’m torn here. It seems like a useful feature but I’m concerned that we’re using the same mechanism that process managers do and there are well known issues with serialization there.

I also wonder (like Yordis) if you couldn’t just manually do this yourself in an event handler.

I don’t want to accept this PR at this point in time. At least not without getting to the bottom of process manager state issues, and that won’t be any time soon.

It is a good idea though @emisilvacab. Thank you for putting the work in.

Reply:
I'm in agreement with @cdegroot and @drteeth. A potentially better approach that already works would be to use an ecto projection (or otherwise create some handler-specific storage of state) so that IO within event handling is made explicit.

Reply:
Unless there is a **strong** technical limitation here, I personally believe we should guide people to create read-models, including a read-model that may feel like a snapshot, make it explicit to the processing of the event